### ISSUE TITLE
ConcurrentLocalContextProvider leaks memory per thread

### ISSUE THREAD

**Issue by matthias-fratz-bsz** [jruby/jruby#8422]
**Environment Information**
  
Provide at least:
* JRuby version: commit 6f9df83e (pom.xml says 9.4.10.0-SNAPSHOT)
* Operating system: Linux 6.1.0-26-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.112-1 (2024-09-30) x86_64 GNU/Linux

**Test Case**

[ThreadMemoryLeak.java](https://github.com/user-attachments/files/17716880/ThreadMemoryLeak.java.txt)

This code creates 10k threads, each of them executing a trivial amount of JRuby code, to simulate a heavily multi-threaded server application churning through threads. It then drops a heap dump for analysis.

Note: This simulates a use case seen in a production Cantaloupe server, which for some reason seems to create and terminate ~4 threads per minute in our case (I don#t understand Jetty's automatic thread pool management there). Because it keeps running for weeks at a time, it churns through tens of thousands of threads, accumulating tens of thousands of ~100kB `LocalContext` objects waiting to be cleaned up in `terminate()`. It eventually slows to a crawl due to memory starvation, or runs out of memory entirely and crashes.

**Expected Behavior**

JRuby creates a `LocalContext` for each thread, holds it in a `ThreadLocal`, and cleans it up when that particular thread has terminated.
The heap dump is a few MB and its size does not scale with the number of terminated threads.

The synthetic example doesn't actually have any concurrent threads; they all are more or less sequential. Thus it should be capable of running with a very small heap, no matter how high you crank the total number of threads that will be created and destroyed (line 36).

**Actual Behavior**

JRuby doesn't remove the `LocalContext` for each thread until the ScriptingContainer itself is disposed. In a long-running server application, that would be "never". In the example, it doesn't happen before the heap dump is written, at which point the JVM is essentially terminating.

The heap dump is hundreds of MB and scales with the number of threads that have been terminated. It is completely dominated by thousands of stale local contexts held by the ScriptingContainer's `ConcurrentLocalContextProvider`.

![biggest-object](https://github.com/user-attachments/assets/ae34c7a7-55d1-4a1e-a46b-492bf213ead2)

![accumulated-objects](https://github.com/user-attachments/assets/ace8d9eb-1434-4a15-b52d-92fa7c094e71)

Setting the number of threads to a high value causes massive heap consumption, and even the synthetic example will eventually run out of memory.

**Probable Cause**

`ConcurrentLocalContextProvider` creates a `LocalContext` for each thread, and only disposes of it in `terminate()`. I do not know the lifecycles of most objects involved, but evidently `termiante()` doesn't get called before JVM termination here. Because these objects are still reachable (via `ConcurrentLocalContextProvider`'s `contextRefs` member), they cannot be garbage-collected.

**Suggested Fix**

Hold a `Reference` (a `PhantomReference` should do, actually) on each thread that a `LocalContext` is created for. When the thread is terminated and becomes unreachable, that `Reference` will show up in its associated `ReferenceQueue`. A background service / cleaner thread can then watch that `ReferenceQueue` and call `remove()` on the relevant `LocalContext` objects. `terminate()` obviously would need to get rid of that service thread, and `remove()` the remaining `LocalContext`s.

Alternatively, the service thread could periodically scan the `contextRefs` array and `remove()` any contexts whose thread has died. This should clean up the context more quickly if anything is still holding a reference to the terminated thread, keeping it form being garbage collected.

Maybe it is also possible to piggy-back scanning onto some other operation, but I'm not sure what the performance impacts of that would be. Or what to do if that operation never happens.

**Comment by matthias-fratz-bsz** [comment]
So... I tried to implement something, and it turns out that `Reference` trickery actually works. The example stays within ~9MB of heap in my case, with Eclipse MAT hilariously reporting `java.lang.Class` as a "leak suspect" ;)

I can offer to make a pull request if you want. It isn't super clean code though; it's more of a PoC.

**Comment by headius** [comment]
@matthias-fratz-bsz Great discovery here, and I'm sorry I didn't reply sooner!

Your original issue relates to leaking LocalContext. Is this BiVariableMap issue a separate concern? If so, I think we should treat them as separate issues, and I would be thrilled if you can start some draft PRs for us to POC these fixes!

The code in this embedding API has been around for a long time and definitely could use some updates and improvements. I would be happy to partner with you to try to clean it up!

**Comment by headius** [comment]
Leaks in the embedding API are not surprising... we should try to clean them up in next minor.

**Comment by headius** [comment]
@matthias-fratz-bsz There's no documentation for how the internals of the ScriptingContainer API works, but there are these docs for general API usage: https://github.com/jruby/jruby/wiki/RedBridge.

Could you put your Reference fix into a PR so we can iterate on it? I think it would help me understand the issue a bit better.

**Comment by matthias-fratz-bsz** [comment]
@headius Here we go, Pull Requst #8483. The actual implementation is in the ThreadLocalContext class, which is now used by both ConcurrentLocalContextProvider and ThreadSafeLocalContextProvider. It actually gets rid of quite a bit of duplicated code.

I actually ended up using the Cleaner API, not Reference directly. This still creates a background Thread that could be an issue with webapp containers, but at least it does so in a way that is common to all use of that official API... so if multiple libraries would do that it's a common fix for e.g. Tomcat.

I will read up on the embedding API a bit more when I find the time. And probably create a new issue / PR for the BiVariable / BiVariableMap leak because it does look like it can be fixed independently.

**Comment by headius** [comment]
Needs a bit more analysis given the introduction of a new cleaner thread. Punting to 9.4.11.0.

**Comment by headius** [comment]
The changes in #8969 are green and waiting on review by @enebo, but I think that will resolve this issue for 9.4.14.0.

**Comment by NC-piercej** [comment]
Given that https://github.com/jruby/jruby/pull/8561 is still open, I don't think that has made it into the JRuby 10 lineage quite yet.

Our app does create a ton of threads and does seem to be affected by this memory leak on JRuby 10 (forcing our ECS tasks to restart as we slowly eventually run out of memory).

**Comment by headius** [comment]
@NC-piercej I'll try to make a JRuby 10.0.3.0 release happen this week, but until then would you be able to test a HEAD build and see if it still leaks? I'd like to make sure we get this right.

https://www.jruby.org/nightly

**Comment by headius** [comment]
@NC-piercej The changes made in #8969 should also be in JRuby 10, but may not be in a 10 *release* as you suspected.
